File: python/leetcode.1116.py
import threading

class ZeroEvenOdd(object):
    def __init__(self, n):
        self.n = n
        self.lock = threading.Lock()
        self.cond = threading.Condition()
        self.index = 0

        
    # printNumber(x) outputs "x", where x is an integer.
    def zero(self, printNumber):
        """
        :type printNumber: method
        :rtype: void
        """
        while 1:
            if self.cond.acquire():
                if self.index >= self.n * 2:
                    self.cond.notify()
                    self.cond.release()
                    break
                elif self.index % 2 == 0:
                    printNumber(0)
                    self.index += 1
                    self.cond.notify()
                else:
                    self.cond.wait()
                self.cond.release()

    def even(self, printNumber):
        """
        :type printNumber: method
        :rtype: void
        """
        while 1:
            if self.cond.acquire():
                if self.index >= self.n * 2:
                    self.cond.notify()
                    self.cond.release()
                    break
                elif self.index % 4 == 3:
                    printNumber((self.index + 1) / 2)
                    self.index += 1
                    self.cond.notify()
                else:
                    self.cond.wait()
                self.cond.release()
        
    def odd(self, printNumber):
        """
        :type printNumber: method
        :rtype: void
        """
        while 1:
            if self.cond.acquire():
                if self.index >= self.n * 2:
                    self.cond.notify()
                    self.cond.release()
                    break
                elif self.index % 4 == 1:
                    printNumber((self.index + 1) / 2)
                    self.index += 1
                    self.cond.notify()
                else:
                    self.cond.wait()
                self.cond.release()

File: cpp/leetcode.1116.cpp
class ZeroEvenOdd {
private:
    int n;
    std::atomic<int> index;
    std::mutex mtx;
    std::condition_variable cv;
    
public:
    ZeroEvenOdd(int n) {
        this->n = n;
        this->index.store(0);
    }

    // printNumber(x) outputs "x", where x is an integer.
    void zero(function<void(int)> printNumber) {
        std::unique_lock<std::mutex> uni_lock(mtx);
        while (1)
        {
            cv.wait(uni_lock, [&](){return index.load() >= 2 * this->n || index.load() % 2 == 0;});
            if(index.load() >= 2 * this->n) return;
            if(index.load() % 2 == 0){
                printNumber(0);
                index++;
            }
            cv.notify_all();
        }
    }

    void even(function<void(int)> printNumber) {
        std::unique_lock<std::mutex> uni_lock(mtx);
        while (1)
        {
            cv.wait(uni_lock, [&](){return index.load() >= 2 * this->n || index.load() % 4 == 3;});
            if(index.load() >= 2 * this->n) return;
            if(index.load() % 4 == 3){
                printNumber((index.load() + 1) / 2);
                index++;
            }
            cv.notify_all();
        }
    }

    void odd(function<void(int)> printNumber) {
        std::unique_lock<std::mutex> uni_lock(mtx);
        while (1)
        {
            cv.wait(uni_lock, [&](){return index.load() >= 2 * this->n || index.load() % 4 == 1;});
            if(index.load() >= 2 * this->n) return;
            if(index.load() % 4 == 1){
                printNumber((index.load() + 1) / 2);
                index++;
            }
            cv.notify_all();
        }
    }
};